server.go

```diff
@@ -673,44 +669,108 @@ func (s *Server) applyCreateShardIfNotExists(m *messaging.Message) (err error) {
 		return ErrRetentionPolicyNotFound
 	}
 
-	// If we can match to an existing shard date range then just ignore request.
-	for _, sh := range rp.Shards {
-		if timeBetweenInclusive(c.Timestamp, sh.StartTime, sh.EndTime) {
-			return nil
-		}
+	// If we can match to an existing shard group date range then just ignore request.
+	if g := rp.shardGroupByTimestamp(c.Timestamp); g != nil {
+		return nil
 	}
 
 	// If no shards match then create a new one.
-	sh := newShard()
-	sh.ID = m.Index
-	sh.StartTime = c.Timestamp.Truncate(rp.Duration).UTC()
-	sh.EndTime = sh.StartTime.Add(rp.Duration).UTC()
+	g := newShardGroup()
+	g.StartTime = c.Timestamp.Truncate(rp.Duration).UTC()
+	g.EndTime = g.StartTime.Add(rp.Duration).UTC()
 
-	// Open shard.
-	if err := sh.open(s.shardPath(sh.ID)); err != nil {
-		panic("unable to open shard: " + err.Error())
+	// Sort nodes so they're consistently assigned to the shards.
+	nodes := make([]*DataNode, 0, len(s.dataNodes))
+	for _, n := range s.dataNodes {
+		nodes = append(nodes, n)
+	}
+	sort.Sort(dataNodes(nodes))
+
+	// Require at least one replica but no more replicas than nodes.
+	replicaN := int(rp.ReplicaN)
+	if replicaN == 0 {
```

Comment:
How could this happen? Shouldn't the action to create the replica block this from ever happening?